### DISCUSSION TITLE
How does TLS client get the DH parameter during handshake stage

### DISCUSSION THREAD

**Discussion by gx2955** [openssl/openssl#21480] [Q&A] [answered]
As TLS client, I want to reject specific server DH parameter at handshake stage. Ex. If server provide MODP-3072 in handshake stage, I want to terminate such connection.

I check as many sources as I could. I do not find any public OpenSSL API support such function.

The closest thing I found is SSL_CTX_set_tmp_dh_callback(), 1. but according to [munu](https://linux.die.net/man/3/ssl_ctx_set_tmp_dh_callback), it is for server only. 2. I test it with sample code, and it is true that client side cannot trigger callback. 3. Even if it works, it just return pkey instead of DH.

**Comment by t8m** [comment, marked as answer]
You cannot do this with TLS-1.2 unfortunately. With TLS-1.3 you can select which groups are supported for key exchange - both DH and ECDH groups are configured with the same Groups entry as described in https://www.openssl.org/docs/man3.0/man3/SSL_CONF_cmd.html

If you want, for TLS-1.2 you can completely disable the DH support and keep only the ECDH support by using `-kDHE:-kDHEPSK` in the cipher string.
